onnxruntime/core/optimizer/reshape_fusion.cc

```diff
@@ -72,6 +79,136 @@ bool ReshapeFusion::Fuse_Subgraph2(Graph& graph, const NodeArg& root_input, cons
+ * to make sure the graph produces output with exactly one element.
+ */
+bool ReshapeFusion::Match_One_Element_Output_Subgraph_2(Graph& graph, const NodeArg& root_input, const Node& cur_node,
+                                                        int index, const logging::Logger& logger) {
+  std::vector<graph_utils::EdgeEndToMatch> parent_path{
+      {0, index, "Squeeze", {1, 11}, kOnnxDomain},
+      {0, 0, "Slice", {1, 11}, kOnnxDomain},
+      {0, 0, "Shape", {1}, kOnnxDomain}};
+  std::vector<const Node::EdgeEnd*> edges;
+  if (graph_utils::FindPath(cur_node, true, parent_path, edges, logger)) {
+    const Node& slice = edges[1]->GetNode();
+    const Node& shape = edges[2]->GetNode();
+
+    const NodeArg& shape_input = *(shape.InputDefs()[0]);
+    if (shape_input.Name() != root_input.Name()) {
+      return false;
+    }
+
+    // Check if Slice op slices 1d array (result of shape) to one element.
+    std::vector<int64_t> slice_inputs;
+    if (optimizer_utils::AppendTensorFromInitializer(graph, *(slice.InputDefs()[1]), slice_inputs, true) &&
+        optimizer_utils::AppendTensorFromInitializer(graph, *(slice.InputDefs()[2]), slice_inputs, true)) {
+      const int64_t slice_start = slice_inputs[0];
+      const int64_t slice_end = slice_inputs[1];
+      if (!(slice_end >= INT_MAX && slice_start == -1) && abs(slice_end - slice_start) != 1) {
+        return false;
+      }
+      return true;
+    }
+  }
+
+  return false;
+}
+
+/**
+ * Check if the i-th input of the current node contains exactly one number.
+ * An input has exactly one element if it's a constant value or it has an inferred shape of [1].
+ */
+bool ReshapeFusion::Is_One_Element_Input(Graph& graph, const Node& cur_node, int index) {
+  const NodeArg* cur_node_arg = cur_node.InputDefs()[index];
+  // Check if the i-th argument is constant, if it is then it's an eligible one element input.
+  if (graph_utils::NodeArgIsConstant(graph, *cur_node_arg)) {
+    std::cout << "nodearg is constant!!!" << std::endl;
+    std::cout << cur_node_arg->Shape() << std::endl;
+    return true;
+  }
+  // Check if the i-th argument has an inferred shape.
+  const auto* input_shape = cur_node_arg->Shape();
+  if (!input_shape) {
+    // We need shape to be able to be certain of number of elements
+    // Can't proceed with fusion
+    return false;
+  }
+
+  // Check if number of elements in this input to perform binary operation is 1
+  if (utils::GetTensorShapeFromTensorShapeProto(*cur_node_arg->Shape()).Size() != 1) {
+    // Some dim values may be > 1 or some dim values may be missing
+    // Can't proceed with fusion
+    return false;
+  }
+  return true;
+}
+
+/**
+ * Search all known patterns of one element subgraphs, which include - 
+ * 1. A concat input with inferred shape that can only contain one element. 
+ * 2. [root] -> Shape -> Gather(any 1d indice) -> Unsqueeze -> [Concat]
+ * 3. [root] -> Shape -> Slice (slice to one element) -> Squeeze -> [Concat]
```

Comment:
Not sure 3 would be possible - Shape by definition produces a 1D tensor - slicing values should keep it at 1D and squeeze would basically remove the rank (if the shape of slice's output is [1]) and thus it would become a scalar. We can't concat a scalar in the next step can we ?

Reply:
Right. Removed.